### PR TITLE
Improvements to a wide range of areas

### PR DESCRIPTION
**Pull request by xBelladonna**
Hi again! I've been pretty sick for a while but I'm back now with another doozy where I changed... everything.

Biggest changes to mention are:
- Extension of templating system that replaces prompt prefix/suffix and also hardcoded square-bracket username schema
  - Added many different config options to enable granular control (see example-config.yml):
    - User display names
    - Bot display name
    - Instruct templates (system token, user token, model token, BOS token, EOS token, etc)
- Improvements to date/time templating
  - Proper consideration of timezone within container (through environment variable TZ)
- Small improvements to image generator
  - Fix timeout hardcoding (actually use setting value)
  - Remove trailing commas from image prompt

Good luck reviewing this, it's a fair bit all at once. Let me know if you have any questions or want to discuss anything.

### CODE REVIEW

**Comment by jmoney7823956789378** [comment]
@xBelladonna welcome back. These are definitely improvements in my eyes, but this branch doesnt seem to run successfully quite yet. 
```
root@chatbot-dev:~# oobabot -c staging.yml 
2024-03-31 23:07:54,088  INFO Starting oobabot, core version 0.2.3
2024-03-31 23:07:54,089 ERROR Unhandled exception
Traceback (most recent call last):
  File "/usr/local/bin/oobabot", line 8, in <module>
    sys.exit(main())
  File "/root/staging/src/oobabot/oobabot.py", line 268, in main
    run_cli()
  File "/root/staging/src/oobabot/oobabot.py", line 264, in run_cli
    oobabot.start()
  File "/root/staging/src/oobabot/oobabot.py", line 93, in start
    self.runtime = runtime.Runtime(self.settings)
  File "/root/staging/src/oobabot/runtime.py", line 52, in __init__
    self.ooba_client = ooba_client.OobaClient(
TypeError: OobaClient.__init__() got an unexpected keyword argument 'template_store'
```

**Comment by xBelladonna** [comment]
Ah god, thanks for catching that. Sleep deprivation strikes again. I'm navigating "branch hell" in order to run a custom head with some commits I don't deem appropriate for the main branch and I must have gotten *something* right in that one because it's working fine. Let me see if I can backport the success.

**Comment by xBelladonna** [comment]
@jmoney7823956789378 would you be able to test again? I *think* I've fixed the problem.

**Comment by xBelladonna** [comment]
As an aside, if you wanted to check out what I was doing that messed this up, it was the latest commit on the `prevent-impersonation` branch (that should be fixed now as well) which goes a step further with stop sequences and adds present users' display names (templated according to the format) to the stop sequences at runtime, so the model will immediately cease generation if it attempts to impersonate another existing user. I have this as a setup because an instance of the bot I'm running is designed for creative roleplay and regularly needs to speak as non-existing people, but not users that are actually present.

There's also some really useful commands I added in the `new-commands` branch that I'll probably PR into this repo as well, if you wanted to have a look, I'm just not sure exactly what you do and don't want in your main branch, heh.

**Comment by AlanMW** [comment]
> There's also some really useful commands I added in the `new-commands` branch that I'll probably PR into this repo as well, if you wanted to have a look, I'm just not sure exactly what you do and don't want in your main branch, heh.

Dumb question, but is that a private branch or something, I'm interested to see what you are working on.

Edit: Nevermind, [i found it](https://github.com/xBelladonna/oobabot/tree/new-commands) of course right after asking...

**Comment by jmoney7823956789378** [comment]
> There's also some really useful commands I added in the `new-commands` branch that I'll probably PR into this repo as well, if you wanted to have a look, I'm just not sure exactly what you do and don't want in your main branch, heh.

I took a look a this branch and it looks great! I do prefer /lobotomize over /neuralyze though...
Give me some time to test alllll this cool stuff completely, but for now:

improvements branch:
failure to start
```
Traceback (most recent call last):
  File "/usr/local/bin/oobabot", line 8, in <module>
    sys.exit(main())
  File "/root/staging/src/oobabot/oobabot.py", line 268, in main
    run_cli()
  File "/root/staging/src/oobabot/oobabot.py", line 264, in run_cli
    oobabot.start()
  File "/root/staging/src/oobabot/oobabot.py", line 93, in start
    self.runtime = runtime.Runtime(self.settings)
  File "/root/staging/src/oobabot/runtime.py", line 92, in __init__
    self.prompt_generator = prompt_generator.PromptGenerator(
  File "/root/staging/src/oobabot/prompt_generator.py", line 76, in __init__
    self.bot_prompt_block = self.template_store.format(
  File "/root/staging/src/oobabot/templates.py", line 387, in format
    return self.templates[template_name].format(format_args)
  File "/root/staging/src/oobabot/templates.py", line 415, in format
    return self.template.format(**format_args)
KeyError: 'BOT_SEQUENCE_PREFIX'
```

new-commands branch: (this branch also does not support the prefixes)

failure on catching immersion breaking lines 

image
```
[Reisen]: Ummmm... what were we talking about?
[stoop poops]: hey reisen! tell me what's going on in eintei today[Reisen]:
 Nothing much, it's a pretty quiet day.
[Reisen]: I was just eating some bamboo shoots when you came in and started bothering me :< [Reisen]: Oh! But the flowers2024-04-01 23:56:58,729 ERROR Exception while running <coroutine object DiscordBot._send_text_response_in_channel at 0x730cfcf4c6d0> response: not enough values to unpack (expected 2, got 1)
Stack (most recent call last):
  File "/usr/local/bin/oobabot", line 8, in <module>
    sys.exit(main())
  File "/root/staging/src/oobabot/oobabot.py", line 268, in main
    run_cli()
  File "/root/staging/src/oobabot/oobabot.py", line 264, in run_cli
    oobabot.start()
  File "/root/staging/src/oobabot/oobabot.py", line 100, in start
    asyncio.run(self.runtime.run())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/usr/lib/python3.10/asyncio/base_events.py", line 1909, in _run_once
    handle._run()
  File "/usr/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/local/lib/python3.10/dist-packages/discord/client.py", line 441, in _run_event
    await coro(*args, **kwargs)
  File "/root/staging/src/oobabot/discord_bot.py", line 245, in on_message
    await self._handle_response(
  File "/root/staging/src/oobabot/discord_bot.py", line 311, in _handle_response
    fancy_logger.get().error(
2024-04-01 23:56:58,729 ERROR Ignoring exception in on_message
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/discord/client.py", line 441, in _run_event
    await coro(*args, **kwargs)
  File "/root/staging/src/oobabot/discord_bot.py", line 245, in on_message
    await self._handle_response(
  File "/root/staging/src/oobabot/discord_bot.py", line 316, in _handle_response
    raise task.exception()
  File "/root/staging/src/oobabot/discord_bot.py", line 479, in _send_text_response_in_channel
    last_sent_message = await self._render_streaming_response(
  File "/root/staging/src/oobabot/discord_bot.py", line 616, in _render_streaming_response
    (response, abort_response) = self._filter_immersion_breaking_lines(response)
  File "/root/staging/src/oobabot/discord_bot.py", line 701, in _filter_immersion_breaking_lines
    username, remaining_text = match.groups()
ValueError: not enough values to unpack (expected 2, got 1)
```

Demo'd using this config. removed the suffix stuff when testing on `new-commands` since that feature is not present there.
[aughhh.txt](https://github.com/jmoney7823956789378/oobabot/files/14829669/aughhh.txt)

**Comment by jmoney7823956789378** [comment]
I still think usernames need a unique identifier (likely via symbols or maybe even old school IRC tags, anything that isnt likely to show up in normal usage) before/around the name, so that impersonation can be reliably caught. Adding the current list of usernames to the filter is good for catching those events involving those users, but if the bot comes up with "user2: hi guys" and then replies with "user6: hi user2" (non-existent users) and so on, it will never end :P 
I know your usage is a little different than mine (my bot only belongs to a single persona), so I'm trying to reach a middle ground between utility and creativity.

**Comment by xBelladonna** [comment]
Waugh... I've got a lot of work to do, thanks for catching those bugs.

I want to be clear that the other branches I mentioned are not designed to run and I haven't tested doing so, they're basically used as a place to store commits for me to cherry-pick/rebase as I work. I've got them all merged in non-Euclidean ways on a custom branch I don't publish. That runs, but it changes too much (in ways that need to be force-pushed) to really publish.

It also seems my bot doesn't fuck up enough to really test the immersion breaking filter well, so I appreciate those tracebacks.

With regards to your last comment, that's exactly what the extension of the template system in this PR is supposed to enable. Just set the config like:
```yaml
user_name: "[{NAME}]"
```
and you should have the square bracket schema like the previous implementation provides, only with the flexibility to make any schema you want in the config. The code ~~will~~ should also use the structure of this template in the immersion breaking filter.

**Comment by xBelladonna** [comment]
I've made some progress reshuffling commits around on branches and force-pushing the various branches. The `prevent-impersonation` and `new-commands` branches *might* run on their own now, and are better organized in terms of commit structure, enabling easier merging/rebasing/cherry-picking. I've also made some more changes on this branch that should fix the exceptions you raised above.

If you want to test all the potential changes I have queued up in those branches, I suggest the following:
1. Start on this branch: `git checkout improvements`
2. Create a new temporary branch: `git branch custom && git checkout custom`
3. Merge `prevent-impersonation`: `git merge prevent-impersonation`
4. Merge `new-commands`: `git merge new-commands`

**Comment by xBelladonna** [comment]
@jmoney7823956789378 looked over your attached example config again and hopefully unfucked the immersion breaking filter. I have edited my previous message containing the example config section to reflect proper usage like in your config. One of these days I should actually document breaking config changes I make....

**Comment by jmoney7823956789378** [comment]
> One of these days I should actually document breaking config changes I make....

Hahaha don't worry to much about it, I've got the feeling oobabot is still in it's "adolescence", and I hope to get it running even better, maybe finally adding CHAT api endpoint support... which would solve prompt prefix hell, since the chat api endpoint will smartly apply formatting as needed.
For now, I tested out both the improvements branch and attempted to branch-stack improvements+prevent-impersonation+new-commands, but as it would turn out, I still recieve the same keyerror as before (on both attempts):
```
Traceback (most recent call last):
  File "/usr/local/bin/oobabot", line 8, in <module>
    sys.exit(main())
  File "/root/staging/src/oobabot/oobabot.py", line 268, in main
    run_cli()
  File "/root/staging/src/oobabot/oobabot.py", line 264, in run_cli
    oobabot.start()
  File "/root/staging/src/oobabot/oobabot.py", line 93, in start
    self.runtime = runtime.Runtime(self.settings)
  File "/root/staging/src/oobabot/runtime.py", line 92, in __init__
    self.prompt_generator = prompt_generator.PromptGenerator(
  File "/root/staging/src/oobabot/prompt_generator.py", line 76, in __init__
    self.bot_prompt_block = self.template_store.format(
  File "/root/staging/src/oobabot/templates.py", line 387, in format
    return self.templates[template_name].format(format_args)
  File "/root/staging/src/oobabot/templates.py", line 415, in format
    return self.template.format(**format_args)
KeyError: 'BOT_SEQUENCE_PREFIX'
```
I'm sorry I can't take a closer look quite yet, as my week gets pretty busy outside of weekends, so the least I can do is skim commits and attempt to test it.
Do you have a containerized/sandbox environment to test these builds in?

**Comment by xBelladonna** [comment]
I swear to god git drives me mad. I do have a containerized build/test/production system and I can promise the bot seems to run on my custom hell-merge branch, so I just have to figure out how to replicate that success, against all odds it seems.

No problems about being busy, I know how that is! I just tend to push stuff and wait until people have time, ~~screwing things up~~ applying more fixes as I notice what I messed up lol

I'll try (again) to merge a working branch that can demo the other changes and just push that so it's ready for you to test.

Also: Chat Completions API support is something I have on the to-do list and would absolutely be a game changer with regards to prompt hell. I still want to support both schemes since they can both be useful, but plan to move forward using the Chat Completions API as the default. One thing at a time, I suppose!

**Comment by xBelladonna** [comment]
Ugh, sorry about the force-push. At least it seems I've got that staging branch set up and working (or running at least) for you to test. It's literally called `staging` and is constructed by starting on the `improvements` branch (this one) and running these, in order:
1. `git merge new-commands`
2. `git merge prevent-impersonation`

**Comment by jmoney7823956789378** [comment]
@xBelladonna looks like this all works. I spun up another test container for your triple-merged branch and it at least started. A little more testing to come later today and saturday, maybe merged then.

**Comment by jmoney7823956789378** [comment]
After another round of testing I found that populating the bot history line like such:
`bot_prompt_history_block: '{BOT_SEQUENCE_PREFIX}{BOT_NAME}: {MESSAGE} {BOT_SEQUENCE_SUFFIX}'`
will result in the error, no matter what is stored in `BOT_SEQUENCE_PREFIX`
I'll keep looking into it.

**Comment by jmoney7823956789378** [comment]
Oops. Well, it looks like the following lines were not edited from the config template:
```
  # Part of the AI response-generation prompt, this is used to render a single line of chat
  # history for users.  A list of these, one for each past user message, will become part of
  # {MESSAGE_HISTORY} and inserted into the main prompt
  # .
  # Allowed tokens: {USER_SEQUENCE_PREFIX}, {USER_SEQUENCE_SUFFIX}, {MESSAGE}, {USER_NAME}
  # .
  #   default:  {USER_NAME}: {MESSAGE}
  user_prompt_history_block: '{USER_NAME}: {MESSAGE}'

  # Part of the AI response-generation prompt, this is used to render a single line of chat
  # history for the bot.  A list of these, one for each past bot message, will become part
  # of {MESSAGE_HISTORY} and inserted into the main prompt
  # .
  # Allowed tokens: {BOT_SEQUENCE_PREFIX}, {BOT_SEQUENCE_SUFFIX}, {MESSAGE}, {BOT_NAME}
  # .
  #   default:  {BOT_NAME}: {MESSAGE}
  bot_prompt_history_block: '{BOT_NAME}: {MESSAGE}'
```
So I had not assumed that the `bot_prompt_history_block` or `user_prompt_history_block` would automatically apply the prefixes and suffixes to these keys since they are listed in the "allowed tokens". If these tokens (`{USER_SEQUENCE_PREFIX}, {USER_SEQUENCE_SUFFIX}` or `{BOT_SEQUENCE_PREFIX}, {BOT_SEQUENCE_SUFFIX}`) are included in their respective history block keys, the bot will fail.

**Comment by xBelladonna** [comment]
@jmoney7823956789378 Oh my god, sorry about that and thanks for catching it. That's what I get for copypaste programming. I've pushed new a new commit to the staging branch that should fix this, if the config is regenerated anyway. Just holding off pushing it to this branch in case I munted it somehow.

**Comment by jmoney7823956789378** [comment]
Don't worry about it, It was a quick fix once i learned how to read again. I re-forked your staging branch to mine, i'm just testing it live before merging to main.

**Comment by jmoney7823956789378** [comment]
@xBelladonna seems to be performing pretty good after just a day of testing. If you'd like, I can close this PR and just force merge from my side, or you can merge your stanging branch into improvements whenever you're ready. If I don't come up with any new features by the end of the weekend I'll just merge to main.
BTW if you're free, the `self-staging` branch is mostly updated (except unsolicited groupchat messages still arent registered), and I think I mis-merged some stuff, as avatar prompting doesn't work.

**Comment by jmoney7823956789378** [comment]
https://github.com/jmoney7823956789378/oobabot/commit/4e25b934d819272db4d2d0e4a493d17caf19f23b
merged.
Will look into properly handling message request queuing, since apparently if two users message at the same time, both requests are processed and somehow merged together. (probably only affects SOME apis... idk)

**Comment by xBelladonna** [comment]
Hold on, what exactly happened here? The branch should have merged with no conflicts and functionality should not have broken. Perhaps I could help fix the head up?

**Comment by xBelladonna** [comment]
Oh I see what's happened. Yep, my code wasn't merged properly and lots of things are broken. Left to continue like this, history will continue diverging unless I open another PR to fix the merge issues. Tthis is fixable but it will require force-pushing.

The idea behind this pull request was NOT to merge the other branches. I was going to open several other pull requests to merge them in an order that would have worked neatly. I actually have many features and bugfixes queued up to open PRs for, I'm just taking a "one thing at a time" kind of attitude.

I think the best way to handle this from here forward is to:
1. Add my fork as a remote on your own local repository: `git remote add xbelladonna https://github.com/xBelladonna/oobabot && git fetch`
2. Checkout your main branch: `git checkout main`
3. Hard reset to commit 479d07f5ed004cce4f64092d5dd33f7386cd7ee1: `git reset --hard 379d07f5ed004cce4f64092d5dd33f7386cd7ee1`
4. *Rebase* the main branch onto my improvements branch: `git rebase xbelladonna improvements`
5. Force push the changes to your origin branch: `git push -f origin main`

If this is done properly, then the latest commit (4e25b934d819272db4d2d0e4a493d17caf19f23b) on the main branch should no longer be needed as it has already been applied on my branch.

I have several other PRs that I am ready to open, however until this is done I cannot do that, as the branch has been.... FUBAR'd for lack of a better word, and it completely breaks all the commits I made in those branches.

**Comment by xBelladonna** [comment]
@jmoney7823956789378 it's been a while since I've written a how-to on performing a "manual" pull request like in the good old days before GitHub when we sent patches through emails.... let me know if you want any more help.

**Comment by jmoney7823956789378** [comment]
> @jmoney7823956789378 it's been a while since I've written a how-to on performing a "manual" pull request like in the good old days before GitHub when we sent patches through emails.... let me know if you want any more help.

Oh dang maybe i'm an idiot but when i checked it said they werent matching up, but when i checked now it says it's just my latest commit that was the only difference? call me crazy.

**Comment by xBelladonna** [comment]
I'm glad everything got sorted out! I'll finalize the code for the other PRs and open them.

**Comment by xBelladonna** [comment]
Oh @jmoney7823956789378 I take it back. I actually checked the branch history for `main` and it is indeed still messed up. I can't merge my new changes (including one that fixes https://github.com/jmoney7823956789378/oobabot/issues/11) until the instructions I posted above are followed. Sorry about this, but I did not intend my other branches to get merged on top of `improvements`. Please let me know if you'd like help fixing the branch timeline. I've thought about adding you on Discord or something and trying to work closer with you so I can better integrate with your workflow and prevent issues like this from occurring. Sometimes I have a lot to communicate on the process of changes I make that are hard to wrap up in a GitHub PR comment.

**Comment by jmoney7823956789378** [comment]
@xBelladonna you can go ahead an add me here then: `jm7583979822903852`
From what I understand, (which is not much), any further changes you make to your `improvements` branch can be merged, since the actual code isn't different between them. Are you trying to sync your fork with mine, to make it easier to keep track of it all?
Also:
```
root@chatbot-dev:~# git clone https://github.com/jmoney7823956789378/oobabot shart
Cloning into 'shart'...
remote: Enumerating objects: 2405, done.
remote: Counting objects: 100% (456/456), done.
remote: Compressing objects: 100% (135/135), done.
remote: Total 2405 (delta 323), reused 363 (delta 287), pack-reused 1949
Receiving objects: 100% (2405/2405), 1.98 MiB | 14.37 MiB/s, done.
Resolving deltas: 100% (1582/1582), done.
root@chatbot-dev:~# cd shart/
root@chatbot-dev:~/shart# git remote add xbelladonna https://github.com/xBelladonna/oobabot && git fetch
root@chatbot-dev:~/shart# git checkout main
Already on 'main'
Your branch is up to date with 'origin/main'.
root@chatbot-dev:~/shart# git reset --hard 479d07f5ed004cce4f64092d5dd33f7386cd7ee1
HEAD is now at 479d07f modified regex a little. i feel dumb
root@chatbot-dev:~/shart# git rebase xbelladonna improvements
fatal: invalid upstream 'xbelladonna'
root@chatbot-dev:~/shart# git rebase            
FETCH_HEAD                     main                           origin/discord.py-self-farts   origin/self-staging 
HEAD                           origin/HEAD                    origin/main                    origin/self-visiontest 
ORIG_HEAD                      origin/dangerousmodelol        origin/olde                    origin/staging
```

**Comment by xBelladonna** [comment]
@jmoney7823956789378 Looking at the logs you posted, it seems it failed to recognize the new remote you added and therefore the rebase didn't work either.... This is a right mess! It's difficult to explain what's gone wrong because git is such a complicated beast, but I'll try my best to illustrate it with images.

### TLDR you merged my testing branches which contained known bugs and now they're in main. We're trying to undo everything before this PR was opened and then merge the `improvements` branch properly so the bugs don't get committed.

Firstly, let's look at my `improvements` branch:
image

We can see a neat commit history joined directly onto the head of your `main` branch (or what it used to be before the frankenmerge).

Now, let's look at your main branch after the merge:
image

We can see that the history is very different, with commits that do not appear in the `improvements` branch. This is because you merged my other *testing* branches, `new-commands` and `prevent-impersonation`, which were NOT ready for deployment, *especially* in that order. Those branches are a work-in-progress. There are known bugs that shouldn't have affected the features intended to be tested, but are bad to have in the main branch. They were intended for a separate pull request *after* the `improvements` branch was merged, once I'd worked out all the bugs.

A proper merge should have the main branch looking exactly like the improvements branch with no extra commits and every commit with the same hash, in the same order.

What I intended to happen was as follows:
1. This PR was merged directly, with no merging of any other branches.
2. I subsequently open two other pull requests for each of the other branches (`new-commands` and `prevent-impersonation`).
3. Those pull requests would be tested and merged in the order they were created.

This would ensure that the branch history would be consistent. This is important because of other changes that I have queued up which *require* that the branches have the same history. That is the entire point of git, version control through an immutable, cryptographically unique history of changes. Because you merged my other two branches in the staging branch that I prepared exclusively for you to test and NOT to merge to main, you created a diverging timeline based on the `staging` branch. It means that all my new work can no longer be merged into your branch. I know that GitHub confirms the merge is possible, but that is the problem. The only thing worse than something that's broken and doesn't work, is something that's broken and *still* works. It would be a lot to post in a GitHub PR comment, as I said before, but I have gone through the code and I can confirm that various bits and pieces are out of place leading to some obvious bugs and other not obvious ones. I'll hit you up on Discord and try to help you sort this out.
